### PR TITLE
Http error callback

### PR DESCRIPTION
**Pull request by mtesseract**
This implements callback support for HTTP errors (#35).
@etorreborre @vitold @amrhassan

### CODE REVIEW

File: tests/Network/Nakadi/Internal/Retry/Test.hs [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  responder <- prepareMockResponse numFailures
+  return HttpBackend
+    { _httpLbs       = mockHttpLbs responder
+    , _responseOpen  = responseOpen
+    , _responseClose = responseClose
+    }
+
+    where mockHttpLbs responder _request = responder
+
+-- | Tests that the callbacked is called exactly numFailures times
```
Comment (etorreborre) [inline tests/Network/Nakadi/Internal/Retry/Test.hs]: s/the callbacked/the callback

File: tests/Network/Nakadi/Internal/Retry/Test.hs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  return $ do
+    current <- atomically $ do
+      n <- readTVar counter
+      modifyTVar counter (+ 1)
+      return n
+    if current >= successNum
+    then return responseSuccess
+    else throwM (HttpExceptionRequest defaultRequest (StatusCodeException responseFailure mempty))
+
+  where responseSuccess = Response
```
Comment (etorreborre) [inline tests/Network/Nakadi/Internal/Retry/Test.hs]: I would just refactor it as
```
defaultResponse = Response
  { responseStatus = status503
  , responseVersion = http11
  , responseHeaders = []
  , responseBody = mempty
  , responseCookieJar = CJ []
  , responseClose' = ResponseClose (pure ())
  }
  responseSuccess = defaultResponse { responseStatus = status200 }
  responseFailure = defaultResponse { responseStatus = status503 }
```

File: src/Network/Nakadi/Internal/Retry.hs [hunk 3]
```diff
@@ -26,37 +27,51 @@ import           Network.HTTP.Types.Status
 import qualified Network.Nakadi.Internal.Lenses  as L
 import           Network.Nakadi.Internal.Types
 
+httpErrorCallback :: MonadIO m => Config -> Request -> HttpException -> RetryStatus -> m ()
```
Comment (etorreborre) [inline src/Network/Nakadi/Internal/Retry.hs]: Maybe a different name here like `call/trigger/invoke httpErrorCallback`? I thought this was defining a callback at first glance. Or just a comment saying so.

**Comment by etorreborre** [comment]
This is great because this is going to give us better visibility in the logs about what's going on.

**Comment by etorreborre** [comment]
👍

**Comment by mtesseract** [comment]
Thanks for your feedback. Will work on this later.

**Comment by mtesseract** [comment]
I think I took care of all your remarks.

**Pull request merged**